Infrastructure for building dynamic files for download asynchronously.

jmchilton opened the pull request:
A more flushed out attempt of #12719. Implements https://github.com/galaxyproject/galaxy/issues/12715.

Three big commits:
- The first adds the backend component and APIs for serving files.
- The second implements sets up infrastructure for using it to generate downloads - including PDF exports and collection downloads.
- The third replaces the export PDF button in the Markdown UI with a variant that will use this service if tasks are enabled. 

All this is coming from the deferred data branch where this mechanism is used for new kinds of exports from Galaxy.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

jmchilton committed:
Short term storage infrastructure.

jmchilton committed:
Update Markdown UI to use STS for PDF export if tasks are enabled.

bgruening commented on lib/galaxy/app.py:
```diff
@@ -491,6 +498,21 @@ def __init__(self, configure_logging=True, **kwargs):
 
+        # Setup infrastructure for short term storage manager.
+        short_term_storage_config_kwds: Dict[str, Any] = {}
+        short_term_storage_config_kwds["short_term_storage_directory"] = self.config.short_term_storage_dir
```
I guess it would be better to stay consistent here and use `self.config.short_term_storage_directory`

jmchilton replied:
I think there is some internal consistency here. The STS package and all its internals uses ``short_term_storage_directory`` which I think is "the right thing" to do. I shorten it in the config definition to match the other config values in the schema - but I don't think we should have library functionality reflect that short hand. I could be wrong about that first part though... but I have been trying to design library functionality independent of of big, overloaded god objects and if ``config`` wasn't around - this is the variable name I would use.

bgruening commented on lib/galaxy/config/schemas/config_schema.yml:
```diff
@@ -493,6 +493,34 @@ mapping:
 
+      short_term_storage_dir:
+        type: str
+        default: short_term_web_storage
```
why is here the _web_ included? I assume this infrastructure is very general so we can remove the `_web_`

jmchilton replied:
I mentioned this below - but I think this is intentional un-general. I'm not trying to replace TMPDIR here - I'm trying to have a directory to place downloads in. It is very specific and very web oriented. It is in the web framework package o Galaxy and I retain the web in places such as here where I think they make sense. Is that okay?

bgruening replied:
Yes, thanks for explaining.

bgruening commented on lib/galaxy/config/schemas/config_schema.yml:
```diff
@@ -493,6 +493,34 @@ mapping:
+        type: int
+        required: false
+        desc: |
+          Default duration before short term storage files can be cleaned up.
```
the description is the same as above

jmchilton replied:
I think I fixed this.

bgruening commented on lib/galaxy/exceptions/error_codes.json:
```diff
@@ -4,6 +4,11 @@
+   {
+      "name": "NO_CONTENT_GENERIC",
+      "code": 204001,
+      "message": "Galaxy has not content to associate with this request."
```
```suggestion
      "message": "Galaxy has no content to associate with this request."
```

bgruening commented on lib/galaxy/config/schemas/config_schema.yml:
```diff
@@ -493,6 +493,34 @@ mapping:
+
+      short_term_storage_default_duration:
+        type: int
+        required: false
```
default: 24 * 60 * 60 ?

bgruening replied:
xref: https://github.com/galaxyproject/galaxy/pull/13691/commits/ed0ae25c7def313f3ab6a50c3f1d98c948775909#diff-210670b6e25ae8ad3868b835ef697eef01ba67be29f059373d86ab5256777e33R31

bgruening replied:
maybe add a comment that we recommend to cleanup this directory with an external process, like tmpwatch?

jmchilton replied:
I've revised the description of that directory to make it clear that Galaxy will clean up the directory and so something like tmpwatch is not needed. I also make it clear that this isn't like a TMPDIR replacement and so tmpwatch is probably still useful other places.

bgruening commented on lib/galaxy/web/short_term_storage/__init__.py:
```diff
@@ -0,0 +1,295 @@
+        safe_makedirs(target_directory)
+        duration = duration or self._config.default_storage_duration or DEFAULT_STORAGE_DURATION
+        maximum_storage_duration = self._config.maximum_storage_duration
+        if duration and maximum_storage_duration and duration > maximum_storage_duration:
```
duration = min(duration, maximum_storage_duration) ?

jmchilton replied:
Those things are nullable...

```
>>> min(1, None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'NoneType' and 'int'
```

bgruening commented on test/integration/test_config_defaults.py:
```diff
@@ -126,6 +127,7 @@
     "shed_tool_config_file": "managed_config_dir",
     "shed_tool_data_path": "tool_data_path",
     "shed_tool_data_table_config": "managed_config_dir",
+    "short_term_storage_directory": "cache_dir",
```
what is the difference between `short_term_storage_dir` and `short_term_storage_directory`

jmchilton replied:
Dropped this config test change - I think this was a no-op.

jmchilton committed:
Cleanup STS PR based on Bjoern's comments.

jmchilton committed:
Implement asynchronous download APIs using STS.

- Allow serving zipped collections using STS.
- Allow serving markdown PDFs using STS.

bgruening reviewed (approved):
Thanks @jmchilton some virtual :cookie: for you :)

nsoranzo commented on lib/galaxy/config/schemas/config_schema.yml:
```diff
@@ -493,6 +493,42 @@ mapping:
+          tasks (in seconds). The default duration is 1 day.
+
+      short_term_storage_maximum_duration:
+        type: int
+        required: false
+        desc: |
+          The maximum duration short term storage files can hosted before they will be marked for
+          clean up. By default there is no default.
```
@jmchilton Not having a maximum is breaking `make config-rebuild`:

Traceback (most recent call last):
  File "lib/galaxy/config/config_manage.py", line 558, in <module>
    main()
  File "lib/galaxy/config/config_manage.py", line 244, in main
    action_func(args, app_desc)
  File "lib/galaxy/config/config_manage.py", line 454, in _build_sample_yaml
    _write_sample_section(args, f, app_desc.app_name, schema)
  File "lib/galaxy/config/config_manage.py", line 490, in _write_sample_section
    _write_option(args, f, key, option_value, as_comment=as_comment)
  File "lib/galaxy/config/config_manage.py", line 504, in _write_option
    option, value = _parse_option_value(option_value)
  File "lib/galaxy/config/config_manage.py", line 527, in _parse_option_value
    raise Exception(f"Failed to parse value for {option}, expected int got None")
Exception: Failed to parse value for {'type': 'int', 'unknown_option': False, 'default': None, 'desc': 'The maximum duration short term storage files can hosted before they will be marked for\nclean up. By default there is no default.\n', 'required': False}, expected int got None
make: *** [Makefile:101: config-rebuild] Error 1

jmchilton replied:
I missed this comment - I am sorry @nsoranzo
